bfvmm/src/vmcs/src/bitmap.cpp

```diff
@@ -0,0 +1,50 @@
+#include <vmcs/bitmap.h>
+
+bitmap::bitmap(uint32_t num_bits)
+{
+    m_length = num_bits / 8;
+
+    if (num_bits % 8)
+    {
+        m_length++;
+    }
+
+    m_bitmap = new uint8_t[m_length];
+
+    for (uint32_t i = 0; i < num_bits; i++)
+    {
+        set_bit(i);
+        clear_bit(i);
+    }
+}
+
+uint8_t *bitmap::address()
+{
+    return m_bitmap;
+}
+
+void bitmap::set_bit(uint32_t n)
+{
+    if ((n / 8) > m_length) return;
+
+    m_bitmap[n / 8] |= (1 << (n % 8));
+}
+
+void bitmap::clear_bit(uint32_t n)
+{
+    if ((n / 8) > m_length) return;
+
+    m_bitmap[n / 8] &= ~(1 << (n % 8));
+}
+
+bool bitmap::bit(uint32_t n)
+{
+    if ((n / 8) > m_length) return false;
```

Comment:
The division indexes the byte within which the bits lie. So 2 and 3 both lie in the first byte (byte 0). Note taken on using division, will explicitly use the shift operator instead so the compiler can't get stupid with power of 2 division.